ipset.go

```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
+func (s *IPSetBuilder) removeBuilder(b *IPSetBuilder) {
+	b.normalize()
+	for _, r := range b.in {
+		s.RemoveRange(r)
+	}
+}
+
+// Complement updates s to contain the complement of its current
+// contents.
+func (s *IPSetBuilder) Complement() {
+	s.normalize()
+	s.out = s.in
+	s.in = []IPRange{
+		IPPrefix{IP: IPv4(0, 0, 0, 0), Bits: 0}.Range(),
+		IPPrefix{IP: IPv6Unspecified(), Bits: 0}.Range(),
+	}
+}
+
+// Intersect updates s to the set intersection of s and b.
+func (s *IPSetBuilder) Intersect(b *IPSet) {
+	var o IPSetBuilder
+	o.Complement()
+	o.RemoveSet(b)
+	s.removeBuilder(&o)
+}
+
+func discardf(format string, args ...interface{}) {}
+
+// debugf is reassigned by tests.
+var debugf = discardf
+
+func (s *IPSetBuilder) IPSet() *IPSet {
+	s.normalize()
+	return &IPSet{
+		rr: append([]IPRange{}, s.in...),
+	}
+}
+
+// IPSet represents a set of IP addresses.
+//
+// The zero value is a valid value representing a set of no IPs.
+type IPSet struct {
+	rr []IPRange // Sorted by IPRange.less
+}
+
+// Clone returns a copy of s that shares no memory with s.
+func (s *IPSet) Clone() *IPSet {
+	return &IPSet{
+		rr: append([]IPRange{}, s.rr...),
+	}
+}
+
+// Ranges returns the minimum and sorted set of IP
+// ranges that covers s.
+func (s *IPSet) Ranges() []IPRange {
+	return append([]IPRange{}, s.rr...)
 }
 
 // Prefixes returns the minimum and sorted set of IP prefixes
 // that covers s.
-// returning a new slice of prefixes that covers all of the given 'add'
-// prefixes with all the 'remove' prefixes removed.
 func (s *IPSet) Prefixes() []IPPrefix {
-	var out []IPPrefix
-	for _, r := range s.Ranges() {
+	out := make([]IPPrefix, 0, len(s.rr))
+	for _, r := range s.rr {
 		out = append(out, r.Prefixes()...)
 	}
 	return out
 }
 
-// ContainsFunc returns a func that reports whether an IP is in s.
-// The returned func operates on a copy of s, so s may be mutated
-// later.
-func (s *IPSet) ContainsFunc() (contains func(IP) bool) {
-	rv := s.Ranges()
-	// TODO(bradfitz): build a faster data structure with
-	// with s.Prefixes()?
-	return func(ip IP) bool {
-		i := sort.Search(len(rv), func(i int) bool {
-			return ip.Less(rv[i].From)
-		})
-		if i == 0 {
+// Equal returns whether s and o represent the same set of IP
+// addresses.
+func (s *IPSet) Equal(o *IPSet) bool {
+	if len(s.rr) != len(o.rr) {
+		return false
+	}
+	for i := range s.rr {
+		if s.rr[i] != o.rr[i] {
 			return false
 		}
-		i--
-		return rv[i].contains(ip)
 	}
+	return true
 }
 
-// Equal reports whether s contains exactly the same IPs as t.
-func (s *IPSet) Equal(t *IPSet) bool {
-	sr := s.Ranges()
-	tr := t.Ranges()
-	if len(sr) != len(tr) {
-		return false
+// Contains returns whether ip is in s.
+func (s *IPSet) Contains(ip IP) bool {
+	for _, r := range s.rr {
+		if r.Contains(ip) {
+			return true
+		}
 	}
-	for i := range sr {
-		if sr[i] != tr[i] {
-			return false
+	return false
+}
+
+// ContainsRange returns whether all IPs in o are in s.
```

Comment:
"reports whether" here and elsewhere

Reply:
Can we use `r` for consistency? We can pick a different local variable name (sr? x?); this one is more important.